astropy/cosmology/io.py

```diff
@@ -0,0 +1,351 @@
+    instantiated. The cosmology class may be specified by name in a column
+    titled ``cosmology`` (case insensitive). As a classmethod ``read`` may
+    be called from any subclass of `~astropy.cosmology.Cosmology`. In that case
+    the cosmology-specifying column is not required. If present, the class will
+    be checked for compatibility (being a subclass) with the calling class.
+    Like the base class, any Cosmology class's ``read`` method can instantiate
+    any of its subclasses.
+
+    The other parameters are passed to the init signature. Parameters not in the
+    signature will be moved to the metadata, except if the key is already
+    present, in which case a `TypeError` will be raised.
+
+    metadata can be included as a column -- called "meta" -- containing a
+    Mapping, like a `dict`. Alternatively, for file formats like ECSV that can
+    hold metadata, any key of same name as the cosmology instance (held in the
+    column "name") with a Mapping value will also be taken as metadata. If both
+    exist, they will be merged, with the column taking higher priority.
+
+    Many formats do not allow unit information, so make sure the values are in
+    the correct units for initializing the cosmology. For this reason ECSV
+    is the recommended file format for Cosmologies.
+
+    Examples
+    --------
+    TODO
+
+    Raises
+    ------
+    TypeError
+        If a parameter is not in the init signature and already has a value in
+        the metadata.
+    """
+
+    def __init__(self, instance, cls):
+        super().__init__(instance, cls, "read")
+
+    def __call__(self, *args, index=None, **kwargs):
+        # shortcuts for determining format
+        format = kwargs.get("format", None)
+        fmt = str(format).lower()
+        fend = str(args[0]).lower().endswith
+
+        # need to determine the format: is it JSON or ECSV?
+        # JSON?
+        if (fmt == "json") or (format is None and fend(".json")):
+            cosmos = self._read_json(*args, key=index, **kwargs)
+        # ECSV?
+        elif (fmt == "ascii.ecsv") or (format is None and fend(".ecsv")):
+            table = QTable.read(*args, **kwargs)
+            cosmos = self.from_table(table, index=index)
+        else:
+            raise ValueError(
+                "`format` must be JSON or 'ascii.ecsv' or None. "
+                "If format is None, `file` must have suffix '.json' or '.ecsv'."
+            )
+
+        return cosmos
+
+    def _read_json(self, file, key=None, **kwargs):
+        # read file, from path-like or file-like
+        if isinstance(file, (str, bytes, os.PathLike)):  # pathlike
+            with open(file, "r") as read_file:
+                data = read_file.read()
+        else:  # file-like : this also handles errors in dumping
+            data = file.read()
+
+        # parse file
+        mapping = json.loads(data)
+
+        return self.from_mapping(mapping, key=key)
+
+    def from_mapping(self, mapping, key=None):
```

Comment:
> More commonly, objects have from_* as class methods and to_* as normal instance methods.

I was relying on ``read`` to be the class method enforcer. I just checked and it isn't. Isn't this a problem for Table as well?

> In essence it hides them in a lower layer and reduces tab-completion discoverability.

True. I was just trying to avoid polluting the physics namespace with these helper functions. I think my new approach is better. I have separate methods ``from/to_table/mapping`` in ``cosmology/io`` and the methods in ``Cosmology.read/write`` just call those.

Reply:
Now there are discoverable ``from/to_table/mapping``, besides the convenience methods in ``Cosmology.read/write``.